### PR TITLE
New - Auth flow

### PR DESCRIPTION
**Pull request by HarveyPeterForFarRankin**
- Completed auth flow (will need tested lol)
- Added refresh tokens table to db
- Added Routers for auth endpoints & posts endpoints
- index.js only now contains the routers

### CODE REVIEW

File: Backend/Middleware/Authentication/index.ts [hunk 1]
```diff
@@ -1,21 +1,22 @@
...
+const { responseCreator } = require("../../utils");
 
 const validateToken = (req, res, next) => {
+  const response = responseCreator(res);
+  const response403 = response(403);
   //get authorisation header value
   const bearerHeader = req.headers["authorization"];
   if (bearerHeader === undefined) {
-    res.sendStatus(403);
+    return response403("failed to authenticate");
```
Comment (mrjackmclean) [inline Backend/Middleware/Authentication/index.ts:10]: could you not just define authenticateFailRes/response403 = response(403)('failed to authenticate')?  appreciate that you use the response403 in later files w/different message but also you reuse this function the same way twice

File: Backend/Routers/auth.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+router.post(login, async (req, res) => {
+  const { username, password } = req.body;
+  const user = await getUserFromDb(username);
+  const response = responseCreator(res);
+  const response403 = response(403);
+  const response200 = response(200);
+  if (!user) {
+    // user does not exist
+    return response403(`Could not authorise credentials`);
```
Comment (mrjackmclean) [inline Backend/Routers/auth.js:64]: could these strings be extracted to a strings file and then you could reuse the string from the authentication middleware file

File: Backend/date.js [hunk 3]
```diff
@@ -0,0 +1,19 @@
+const pipe = require("ramda/src/pipe");
```
Comment (mrjackmclean) [inline Backend/date.js:1]: i will look at this stuff when i am not on holiday but think i will need explained haha

File: README.md [hunk 4]
```diff
@@ -1,2 +1,15 @@
 # nearbii
+
 An early stage development social media project, involving anonymous posts that are accessible based on location.
+
+# BACKEND
+
+The Backend is using Postgres with Prisma ORM
```
Comment (mrjackmclean) [inline README.md:7]: Why postgres btw?

File: Backend/Middleware/Authentication/index.ts (hunk 1 above)
Comment (mrjackmclean) [inline Backend/Middleware/Authentication/index.ts:10]: forgive me for making this comment without actual booting a single thing up (will do after my holiday) but why 'failed to authenticate' and not {message: 'failed to authenticate'}?

File: Backend/Routers/auth.js (hunk 2 above)
Comment (HarveyPeterForFarRankin) [reply Backend/Routers/auth.js:64]: can dooo

File: Backend/date.js (hunk 3 above)
Comment (HarveyPeterForFarRankin) [reply Backend/date.js:1]: Yeah this is cool stuff

File: README.md (hunk 4 above)
Comment (HarveyPeterForFarRankin) [reply README.md:7]: best realtional db out there

File: Backend/Middleware/Authentication/index.ts (hunk 1 above)
Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts:10]: think i forgot

Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts:10]: Just passing the res as the initial function

Comment (mrjackmclean) [reply Backend/Middleware/Authentication/index.ts]: but i mean could you not just define on this line
--> const failedAuthResponse = response(403)('failed to authenticate') 
and then return it on line 10 and line 16

Comment (mrjackmclean) [reply Backend/Middleware/Authentication/index.ts]: haha fair, not a bad idea tho - lot more readable. could maybe change the the response function to automatically return {message: '<text>'} if the input type is a string?

File: Backend/date.js (hunk 3 above)
Comment (mrjackmclean) [reply Backend/date.js:1]: love it

File: README.md (hunk 4 above)
Comment (mrjackmclean) [reply README.md:7]: yeah we probs need to talk cause they said they were using postgres for a gps thing at work cause it has extra functionality etc but like ............ why use the other ones then

File: Backend/Middleware/Authentication/index.ts (hunk 1 above)
Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts]: that would call the function on that line

Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts]: added message :)

File: README.md (hunk 4 above)
Comment (HarveyPeterForFarRankin) [reply README.md:7]: Postgres is now used over mysql cause it allows us to do a bunch more. can store json etc and large files if need be

File: Backend/Middleware/Authentication/index.ts (hunk 1 above)
Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts]: oh i didnt have it automatically have message as i wanted to allow u to pass any object for a 200 or something if that makes sense

Comment (HarveyPeterForFarRankin) [reply Backend/Middleware/Authentication/index.ts]: Screenshot 2021-12-30 at 18 10 34
for example haha

Comment (mrjackmclean) [reply Backend/Middleware/Authentication/index.ts]: Ah okay - doesn't that mean that the 'return' here is redundant? 
Ps doing this on my phone so can't properly check x

Comment (mrjackmclean) [reply Backend/Middleware/Authentication/index.ts]: Could you not do any object BUT if typeof === string then just use the message? Not super consistent but maybe a bit tidier

**Pull request merged**